engine/source_sym_engine.ml

```diff
@@ -61,122 +20,102 @@ let print_result stree =
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
-    | Conjunction (cst_list) ->
-      bprintf buf
-         "Conjunction:%t%a"
-         (break (ntabs+1))
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
-    | Root cst_list ->
-       bprintf buf
-         "Root=\
-          %t%a"
-         (break (ntabs+1))
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Node (k_cst_list, cst_opt) ->
+      bprintf buf "Node:{\
+                   %a \
+                   %t} Fallback: %a"
+        (bprint_list ~sep:sep
+           (fun buf (k,cst) -> bprintf buf "%t%a -> %t%a"
+               sep
+               bprint_constructor k
+               (break (ntabs+2))
+               (bprint_tree (ntabs+1)) cst))
+        k_cst_list
+        (break ntabs)
+        (fun buf -> begin match cst_opt with 
+             | Some cst -> fun _ -> bprint_tree (ntabs+1) buf cst
+             | None -> fun _ -> bprintf buf "None" 
+           end) buf
   in
   let buf = Buffer.create 42 in
   bprint_tree 0 buf stree;
   BatIO.write_line BatIO.stdout (Buffer.contents buf)
 
-let sym_exec (source: source_program) =
-  let shares_kst ~wildcard_to_none k ((pattern, expr):clause): clause option =
-    match pattern with
-    | Wildcard when wildcard_to_none = false-> Some (Wildcard, expr)
-    | Constructor (k', plist) when k' = k ->
-      if (List.length plist > 1) then
-        Some (Constructor (Tuple, plist), expr)
-      else
-        Some (plist |> List.hd, expr)
-    | _ -> None
-  in
-  let negate_piop = function
-    | EqInt i -> NeqInt i
-    | NeqInt i -> EqInt i
-    | IsK s -> NotK s
-    | NotK s -> IsK s
-    | IsTrue -> IsFalse
-    | IsFalse -> IsTrue
-    | IsTuple -> NotTuple
-    | NotTuple -> IsTuple
-    | IsNil -> NotNil
-    | NotNil -> IsNil
-    | IsCons -> NotCons
-    | NotCons -> IsCons
-    | EqString s -> NeqString s
-    | NeqString s -> EqString s
-  in
-  let piop_of_constructor = function
-    | Variant s -> IsK s
-    | Int i -> EqInt i
-    | Bool b -> if b then IsTrue else IsFalse
-    | String s -> EqString s
-    | Tuple -> IsTuple
-    | Nil -> IsNil
-    | Cons -> IsCons
+type row = pattern list * source_expr
+
+let sym_exec source =
+  let kst_rows_assoc rows : (constructor * row list) list * row option =
+    let hashtbl = Hashtbl.create 42
+    in
+    let with_wildcard = ref []
+    in
+    let relevant_wildcard_opt = function
+      | [] -> None
+      | wildcards -> Some (wildcards |> List.rev |> List.hd)
+    in
+    let map_to_hashtable = 
+      fun ((pattern, expr):row) ->
+        match List.hd pattern with
+        | Constructor (k, plist) ->
+          let binding = match Hashtbl.find_opt hashtbl k with
+            | Some clause_lst -> (plist, expr)::clause_lst
+            | None -> (plist, expr)::[]
+          in
+          Hashtbl.replace hashtbl k binding
+        | Wildcard -> with_wildcard := (pattern, expr)::!with_wildcard
+        | _ -> ()
```

Comment:
The handling of the rest of the pattern list is wrong here: you completely forget about `List.tl pattern`.

I think that you should change the input type of `kst_row_assoc` to take not a `(pattern list * expr) list` but a `(pattern * pattern list * expr) list` (you don't actually need an annotation for that, although you can have one if it helps you). The idea is that you already know, when you enter the function, that the pattern list is non-empty; `pattern * pattern list` represents a non-empty list of patterns.

Then the handling should be something like

```
fun (head, row, expr) ->
  match head with
  | Constructor (k, plist) ->
     let group = try Hashtbl.find hashtbl k with Not_found -> [] in
     let binding = (plist @ row, expr) :: group in
     Hashtbl.replace hahshtbl k binding
```

In the wildcard case, you would not store `(pattern, expr)` but `(rest, expr)`; note that the head pattern (that is `Wildcard`) is removed.

Reply:
The case `_ -> ()` is also fishy. I see two solutions:
- either you expand `Or`-patterns away before you call this function (in the main loop), so we know that there are no `Or`-patterns here, and you can `assert false` in the `Or` case (in any case you should list all possible constructors here)
- or you do the expansion here, by having the function call itself recursively on both arguments (in the same style, it could deal with `As`-patterns by just calling itself recursively on the subpattern); yes, currently this function is anonymous and not recursive